Use EditorFileDialog instead of FileDialog in the project manager

Geometror opened the pull request:
Use EditorFileDialog instead of FileDialog in the project manager (primarily because of the favorites/recent_dir functionality), aiming to accelerate the workflow when frequently creating new projects/dealing with many projects in general. This allows for easier navigation to common folders in which you create many small projects (e.g. for testing/prototyping).

Changes:
- Use EditorFileDialog instead of FileDialog in the project manager
    - Only single files and empty directories can be deleted (for safety/implementation simplicity)
- Add option to disable previews in EditorFileDialog
- Add a proper project_manager_hint to the Engine singleton (removed Main::is_project_manager)
- some refactoring of EditorFileDialog to reduce code duplication and improve code style

![main](https://user-images.githubusercontent.com/50084500/133602798-f719eacd-b823-41d5-8894-8dea84c160c6.gif)

Geometror commented:
Rebased and changed the following:
- Add a warning popup (renamed the mkdirerr AcceptDialog and reuse it) when the path of a favorites entry does not exist anymore and remove it.

akien-mga commented on .godot/.gdignore:
```diff
@@ -0,0 +1 @@
+
```
This was added by mistake.

akien-mga commented on core/config/engine.h:
```diff
@@ -116,6 +117,9 @@ class Engine {
 	void remove_singleton(const StringName &p_name);
 	bool is_singleton_user_created(const StringName &p_name) const;
 
+	_FORCE_INLINE_ void set_project_manager_hint(bool p_enabled) { project_manager_hint = p_enabled; }
+	_FORCE_INLINE_ bool is_project_manager_hint() const { return project_manager_hint; }
```
This should likely be in `#ifdef TOOLS_ENABLED` like the editor hint.

akien-mga commented on editor/dependency_editor.cpp:
```diff
@@ -451,6 +451,8 @@ void DependencyRemoveDialog::show(const Vector<String> &p_folders, const Vector<
 
 	for (int i = 0; i < p_folders.size(); ++i) {
 		String folder = p_folders[i].ends_with("/") ? p_folders[i] : (p_folders[i] + "/");
+		print_line(p_folders[i]);
```
Will be spammy.

akien-mga commented on editor/project_manager.cpp:
```diff
@@ -55,7 +56,8 @@
 #include "servers/navigation_server_3d.h"
 
-static inline String get_project_key_from_path(const String &dir) {
+static inline String
+get_project_key_from_path(const String &dir) {
```
Unexpected line break:
```suggestion
static inline String get_project_key_from_path(const String &dir) {
```

akien-mga commented on editor/project_manager.cpp:
```diff
@@ -924,10 +926,12 @@ class ProjectDialog : public ConfirmationDialog {
-		fdialog_install = memnew(FileDialog);
-		fdialog_install->set_access(FileDialog::ACCESS_FILESYSTEM);
+		fdialog = memnew(EditorFileDialog);
+		fdialog->set_previews_enabled(false); //Crucial, otherwise the Engine crashes.
+		fdialog->set_access(EditorFileDialog::ACCESS_FILESYSTEM);
+		fdialog_install = memnew(EditorFileDialog);
+		fdialog_install->set_previews_enabled(false); //Crucial, otherwise the Engine crashes.
```
```suggestion
		fdialog->set_previews_enabled(false); // Crucial, otherwise the engine crashes.
		fdialog->set_access(EditorFileDialog::ACCESS_FILESYSTEM);
		fdialog_install = memnew(EditorFileDialog);
		fdialog_install->set_previews_enabled(false); // Crucial, otherwise the engine crashes.
```

akien-mga commented on editor/project_manager.cpp:
```diff
@@ -1903,6 +1907,24 @@ void ProjectManager::_notification(int p_what) {
+	return icon_type_cache["ObjectHR"];
+}
+
+void ProjectManager::_build_icon_type_cache(Ref<Theme> p_theme) {
+	List<StringName> tl;
+	StringName ei = "EditorIcons";
+	p_theme->get_icon_list(ei, &tl);
+	for (List<StringName>::Element *E = tl.front(); E; E = E->next()) {
+		if (!ClassDB::class_exists(E->get())) {
+			continue;
+		}
+		icon_type_cache[E->get()] = p_theme->get_icon(E->get(), ei);
+	}
+}
```
```suggestion
void ProjectManager::_build_icon_type_cache(Ref<Theme> p_theme) {
	List<StringName> tl;
	p_theme->get_icon_list(SNAME("EditorIcons"), &tl);
	for (List<StringName>::Element *E = tl.front(); E; E = E->next()) {
		if (!ClassDB::class_exists(E->get())) {
			continue;
		}
		icon_type_cache[E->get()] = p_theme->get_icon(E->get(), SNAME("EditorIcons"));
	}
}
```

akien-mga commented on editor/editor_file_dialog.cpp:
```diff
@@ -1730,6 +1786,11 @@ EditorFileDialog::EditorFileDialog() {
 	remove_dialog = memnew(DependencyRemoveDialog);
 	add_child(remove_dialog);
 
+	project_man_remove_dialog = memnew(ConfirmationDialog);
+	project_man_remove_dialog->set_text(TTR("Remove the selected files? For safety only files and empty directories can be deleted from here. (Cannot be undone.)\nDepending on your filesystem configuration, the files will either be moved to the system trash or deleted permanently."));
+	project_man_remove_dialog->connect("confirmed", callable_mp(this, &EditorFileDialog::_delete_files_project_man));
+	add_child(project_man_remove_dialog);
```
I'm not sure I like the idea to have Project Manager specific code in EditorFileDialog. It's meant to be a general purpose file dialog, it would be better if there were relevant hooks for the Project Manager to configure the behavior that it wants (e.g. signals it can connect its own callbacks to).

Geometror replied:
You're right, this is not how it should be. Fixed accordingly (see comment).

akien-mga commented on editor/editor_file_dialog.cpp:
```diff
@@ -1184,10 +1203,39 @@ void EditorFileDialog::_update_drives(bool p_select) {
+		error_dialog->set_text(TTR("Favorized folder does not exist anymore and will be removed."));
+		error_dialog->popup_centered(Size2(250, 50) * EDSCALE);
+
+		bool res = access == ACCESS_RESOURCES;
```
```suggestion
		bool res = (access == ACCESS_RESOURCES);
```
To be consistent with the style change done below.

akien-mga commented on editor/editor_file_dialog.cpp:
```diff
@@ -1184,10 +1203,39 @@ void EditorFileDialog::_update_drives(bool p_select) {
-	_push_history();
+	Error change_dir_result = dir_access->change_dir(favorites->get_item_metadata(p_idx));
+	if (change_dir_result != OK) {
+		error_dialog->set_text(TTR("Favorized folder does not exist anymore and will be removed."));
```
I'm not sure "Favorized" will be understood / is a common term for this.

aaronfranke replied:
The English term for this would be "Favorited", not "Favorized".

Geometror replied:
Oh, you are right, changed it to "favorited".

Geometror committed:
Use EditorFileDialog instead of FileDialog in the project manager

Geometror commented:
Fixed the addressed issues, resulting in some changes to EditorFileDialog:
- It does no longer contain project manager related code. Dialogs which access the "global" filesystem (meaning not the project directory) now support deletion (only files and empty folders for safety), not only the file dialogs in the project manager (for example, the editor feature import dialog/export GLTF dialog). This option was already there in those dialogs, but resulted in an error since the DependencyRemoveDialog was always used for deletion. 

In addition:
- Some minor refactoring and style improvements (e.g. added method `_update_icons` to reduce code duplication/ removed some comments)

akien-mga commented:
Needs a rebase to solve merge conflicts.

akien-mga commented:
Thanks!

bruvzg commented:
This change is causing Project Manager to crash on exit:

Trace:
```
Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Codes:       0x0000000000000001, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [22513]

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   godot.osx.opt.tools.arm64     	       0x107c8e928 ImageTexture::~ImageTexture() + 48 (texture.cpp:290)
1   godot.osx.opt.tools.arm64     	       0x107c8e924 ImageTexture::~ImageTexture() + 44 (texture.cpp:290)
2   godot.osx.opt.tools.arm64     	       0x10639ddf4 void memdelete<Texture2D>(Texture2D*) + 28 (memory.h:108) [inlined]
3   godot.osx.opt.tools.arm64     	       0x10639ddf4 Ref<Texture2D>::unref() + 48 (ref_counted.h:221) [inlined]
4   godot.osx.opt.tools.arm64     	       0x10639ddf4 Ref<Texture2D>::~Ref() + 48 (ref_counted.h:233) [inlined]
5   godot.osx.opt.tools.arm64     	       0x10639ddf4 Ref<Texture2D>::~Ref() + 48 (ref_counted.h:232) [inlined]
6   godot.osx.opt.tools.arm64     	       0x10639ddf4 KeyValue<String, Ref<Texture2D> >::~KeyValue() + 48 (pair.h:73) [inlined]
7   godot.osx.opt.tools.arm64     	       0x10639ddf4 KeyValue<String, Ref<Texture2D> >::~KeyValue() + 48 (pair.h:73) [inlined]
8   godot.osx.opt.tools.arm64     	       0x10639ddf4 Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element::~Element() + 48 (map.h:50) [inlined]
9   godot.osx.opt.tools.arm64     	       0x10639ddf4 Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element::~Element() + 48 (map.h:50) [inlined]
10  godot.osx.opt.tools.arm64     	       0x10639ddf4 void memdelete_allocator<Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element, DefaultAllocator>(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 64 (memory.h:120)
11  godot.osx.opt.tools.arm64     	       0x10639dd98 Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::_cleanup_tree(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 52 (map.h:571)
12  godot.osx.opt.tools.arm64     	       0x10639dd8c Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::_cleanup_tree(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 40 (map.h:570)
13  godot.osx.opt.tools.arm64     	       0x10639dd8c Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::_cleanup_tree(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 40 (map.h:570)
14  godot.osx.opt.tools.arm64     	       0x10639dd8c Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::_cleanup_tree(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 40 (map.h:570)
15  godot.osx.opt.tools.arm64     	       0x10639dd8c Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::_cleanup_tree(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 40 (map.h:570)
16  godot.osx.opt.tools.arm64     	       0x10639dd8c Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::_cleanup_tree(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 40 (map.h:570)
17  godot.osx.opt.tools.arm64     	       0x10639dd8c Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::_cleanup_tree(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 40 (map.h:570)
18  godot.osx.opt.tools.arm64     	       0x10639dd8c Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::_cleanup_tree(Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::Element*) + 40 (map.h:570)
19  godot.osx.opt.tools.arm64     	       0x106601608 Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::clear() + 20 (map.h:732) [inlined]
20  godot.osx.opt.tools.arm64     	       0x106601608 Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::~Map() + 20 (map.h:749) [inlined]
21  godot.osx.opt.tools.arm64     	       0x106601608 Map<String, Ref<Texture2D>, Comparator<String>, DefaultAllocator>::~Map() + 36 (map.h:748)
22  libsystem_c.dylib             	       0x1a3e747d8 __cxa_finalize_ranges + 460
23  libsystem_c.dylib             	       0x1a3e74580 exit + 44
24  libdyld.dylib                 	       0x1a3f91ea8 dyld4::LibSystemHelpers::exit(int) const + 20
25  dyld                          	       0x10e3a9140 start + 596
```

Seems like something wrong with loading/unloading editor icons.

bruvzg commented:
Also, it seems to completly break editor setting saving/loading, everything is written to the current directory instead of proper global location.
